### ISSUE TITLE
Expose clobberTimes as a flag to ADD instruction

### ISSUE THREAD

**Issue by bjaglin** [docker/docker#4351]
In my scenario, I am `ADD`ing a bunch of libraries packaged as jars, copied over in a `target` directory during build time. I would like that instruction to be cached across builds (unless new libraries are added obviously, which doesn't happen very often).

When `ADD`ing a URL in a Dockerfile, ctime, atime and mtime are not considered for the hash (see https://github.com/dotcloud/docker/pull/2809/files#diff-d8fbf83f530e600c64298c1b18b04d4aR411). It would be interesting to expose a flag controlling that `clobberTimes` for local files/directories, to be able to utilize caches in cases where the metadata of a file doesn't matter. I am thinking about adding a `--ignore-times` or `--clober-times` flag to the `ADD` instruction but I can't see any option in the current set of instructions, so it might be considered an anti-pattern. For the effect of that flag to be completely consistent across image builders, the times should also be reset to a default value (0?), so that the use of this option produces the exact same image no matter what the times on the filesystem of the host building the image are.

I am willing to contribute a patch if there is some interest, and some agreement on the exact naming/semantics of that flag.

Running Docker 0.8.1 on Ubuntu 14.04.

**Comment by jfrazelle** [comment]
@duglin didnt we add mtime

**Comment by duglin** [comment]
@jfrazelle yes mtime is considered part of the cache check.  Once https://github.com/docker/docker/pull/10775 is merged (hint hint :-) ) then we can look at how to add flags like this - its a very popular request.

**Comment by Silex** [comment]
I was asked to merge my issue into this one, my proposal was like so:

``` man 
--cache-strategy timestamp|sha1sum|md5sum

timestamp (default): use the modification time of the file (fast). Only cares about file timestamp.
sha1sum: use sha1sum to determinate if file has changed. Only cares about file contents.
md5sum: use md5sum to determinate if file has changed (slow). Only cares about file contents.
```

**Comment by wrossmck** [comment]
+1 @Silex

**Comment by rflynn** [comment]
+1

**Comment by soupdiver** [comment]
I also like https://github.com/docker/docker/issues/4351#issuecomment-81187546
:+1:

**Comment by UnquietCode** [comment]
I'm digging the idea of a cache strategy flag ([comment](https://github.com/docker/docker/issues/4351#issuecomment-81187546) above) but would suggest that the default be set to use the checksum. Also, supporting and exposing different hash algorithms seems unnecessary and confusing.

Thus:
```
--cache-strategy checksum|timestamp|

checksum: (default) use a hashing algorithm to determinate if file has changed. Only cares about file contents.
timestamp:  use the modification time of the file (fast). Only cares about file timestamp.
```

Where timestamp would probably use `mtime` and could make use of HTTP headers and other sources of timing data.

**Comment by duglin** [comment]
I'm +1 to allowing the user to specify some flag to tell the builder to use checksums only and not timestamps. I think enough people have complained that we need to do something to help them.

I'm -1 to the idea of changing the defaults though.  Unless we can be 100% sure that **no one** is counting on today's behavior (which I don't think we can do), I think changing the default on people would be a mistake. Eg. I can see people using 'touch' to force a cache miss - I think we'd break them.

One thing that's less clear to me what the flag should look like.  `--cache-strategy` is certainly the proper technical description of it, but... is it too technical (geeky)?  Do we expect to have lots of new caching strategies in the future? If so, then that flag make sense. But if not then I'm leaning more towards something like "--ignore-mtime" which is more restrictive but also more user-friendly.

**Comment by dreamcat4** [comment]
Hello. Please make sure the solution works with *not downloading the github releases every time*. That website is very popular souce to `ADD` tarballs from. Thank you.

**Comment by wrossmck** [comment]
@duglin I'd argue that `--cache-strategy` is more user friendly than `--ignore-mtime`, and there are the obvious benefits of not being locked into a specific flip-flop flag. 

+1 for `--cache-strategy`
+1 for not changing the defaults

**Comment by dreamcat4** [comment]
Why not just abbreviate `--cache-strategy` to `--cache` for short? Then:

```Dockerfile
ADD --cache=checksum https://github.com/account/repo/.../archive.tgz archive.tgz
# or
ADD --cache=mtime https://github.com/account/repo/.../archive.tgz archive.tgz
```

Oh wait. Did I mention that I also need that in conjunction with a `--extract` flag all on the same line?

```Dockerfile
ADD --cache=mtime --extract https://github.com/account/repo/.../archive.tgz archive.tgz
# looks better than
ADD --cache-strategy=mtime --extract https://github.com/account/repo/.../archive.tgz archive.tgz
```

Unnecessarily long / verbose arguments like `--cache-strategy` will ultimately make the `Dokerfile` more trouble to read / bit gets scrolled that much extra off the page.

**Comment by j0hnsmith** [comment]
+1 https://github.com/docker/docker/issues/4351#issuecomment-81187546

**Comment by Silex** [comment]
Using `--cache` instead of `--cache-strategy` looks good. I didn't put *that* much thought into it, improvements are always welcome.

My original suggestion was `docker build --cache=md5 -t foo`, but having it tunable for `ADD` or `COPY` statements  looks even more flexible, but I'd still like the global `docker build` flag, because it's rare that you need a flag for one copy and a different one for another.

**Comment by thaJeztah** [comment]
We recently merged a pull-request that makes the cache ignore mtime; https://github.com/docker/docker/pull/12031. After some discussion, it was considered to be a bug that mtime was taken into account to invalidate the cache (https://github.com/docker/docker/pull/12031#issuecomment-101482148).

Unfortunately the PR didn't make it in time for the 1.7 release, but it should be in the 1.8 release.

**This means that with the 1.8 release, mtime will be ignored for local files.** 

For people "mis-using" that behavior to deliberately bust the cache, this can be a breaking change, and a different approach should be taken.

I'm going to close this issue, because the "mtime" bug is now resolved in https://github.com/docker/docker/pull/12031

If there's still a need to have more fine-grained control over cache strategies (as discussed above), please open a new issue.
